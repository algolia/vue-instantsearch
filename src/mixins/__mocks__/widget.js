let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
