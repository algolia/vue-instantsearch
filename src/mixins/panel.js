import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
