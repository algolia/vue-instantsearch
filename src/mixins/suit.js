import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
