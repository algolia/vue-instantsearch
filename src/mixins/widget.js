const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
