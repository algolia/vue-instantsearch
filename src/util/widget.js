export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
