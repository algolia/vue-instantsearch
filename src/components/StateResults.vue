<template>
  <div
    :class="suit()"
    v-if="state && state.state && state.results"
  >
    <slot v-bind="stateResults">
      <p>
        Use this component to have a different layout based on a certain state.
      </p>
      <p>
        Fill in the slot, and get access to the following things on the
        <code>slot-scope</code>:
      </p>
      <pre>results: {{ Object.keys(state.results) }}</pre>
      <pre>state: {{ Object.keys(state.state) }}</pre>
    </slot>
  </div>
</template>

<script>
import { createSuitMixin } from '../mixins/suit';
import { createWidgetMixin } from '../mixins/widget';
import { _objectSpread } from '../util/polyfills';
import connectStateResults from '../connectors/connectStateResults';

export default {
  name: 'AisStateResults',
  mixins: [
    createWidgetMixin({ connector: connectStateResults }),
    createSuitMixin({ name: 'StateResults' }),
  ],
  computed: {
    stateResults() {
      // @MAJOR: replace v-bind="stateResults" with :state="state.state" :results="state.results"
      const { state, results } = this.state;
      return _objectSpread({}, results, { results, state });
    },
  },
};
</script>
