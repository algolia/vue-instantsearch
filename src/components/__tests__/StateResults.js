import { mount } from '@vue/test-utils';
import StateResults from '../StateResults.vue';
import { __setState } from '../../mixins/widget';
jest.mock('../../mixins/widget');

it('renders explanation if no slot is used', () => {
  __setState({
    results: {
      query: 'this is the quer',
      hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
      page: 1,
    },
    state: {
      query: 'this is the query',
    },
  });
  const wrapper = mount(StateResults);
  expect(wrapper.html()).toMatchSnapshot();
});

it("doesn't render if no results", () => {
  __setState({});
  const wrapper = mount(StateResults);
  expect(wrapper.html()).toBeUndefined();
});

it('gives state & results to default slot', () => {
  const results = {
    query: 'q',
    hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
    page: 1,
  };
  const state = {
    query: 'something',
    disjunctiveFacetsRefinements: {},
    page: 1,
  };

  __setState({
    state,
    results,
  });

  mount(StateResults, {
    scopedSlots: {
      default: props => {
        expect(props).toEqual(expect.objectContaining(results));
        expect(props.results).toEqual(results);
        expect(props.state).toEqual(state);
      },
    },
  });
});

it('allows default slot to render whatever they want', () => {
  const results = {
    query: 'hi',
    hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
    page: 1,
  };
  const state = {
    query: 'hi!',
  };
  __setState({
    state,
    results,
  });

  const wrapper = mount(StateResults, {
    scopedSlots: {
      default: `
      <template slot-scope="{ state: { query }, results: { page } }">
        <p v-if="query">
          Query is here, page is {{ page }}
        </p>
        <p v-else>
          There's no query
        </p>
      </template>`,
    },
  });

  expect(wrapper.html()).toMatchInlineSnapshot(`

<div class="ais-StateResults">
  <p>
    Query is here, page is 1
  </p>
</div>

`);
});

it('allows default slot to render whatever they want (truthy query)', () => {
  const results = {
    query: 'hi',
    hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
    page: 1,
  };
  const state = {
    query: 'hi!',
  };
  __setState({
    state,
    results,
  });

  const wrapper = mount(StateResults, {
    scopedSlots: {
      default: `
      <template slot-scope="{ results: { query } }">
        <p v-if="query">
          Query is here
        </p>
        <p v-else>
          There's no query
        </p>
      </template>`,
    },
  });

  expect(wrapper.html()).toMatchInlineSnapshot(`

<div class="ais-StateResults">
  <p>
    Query is here
  </p>
</div>

`);
});

it('allows default slot to render whatever they want (falsy query)', () => {
  const results = {
    query: '',
    hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
    page: 1,
  };
  const state = {
    query: 'hi!',
  };
  __setState({
    state,
    results,
  });

  const wrapper = mount(StateResults, {
    scopedSlots: {
      default: `
      <template slot-scope="{ results: { query } }">
        <p v-if="query">
          Query is here
        </p>
        <p v-else>
          There's no query
        </p>
      </template>`,
    },
  });

  expect(wrapper.html()).toMatchInlineSnapshot(`

<div class="ais-StateResults">
  <p>
    There's no query
  </p>
</div>

`);
});

describe('legacy spread props', () => {
  it('allows default slot to render whatever they want (truthy query)', () => {
    __setState({
      results: {
        query: 'q',
        hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
        page: 1,
      },
      state: {},
    });

    const wrapper = mount(StateResults, {
      scopedSlots: {
        default: `
        <template slot-scope="{ query }">
          <p v-if="query">
            Query is here
          </p>
          <p v-else>
            There's no query
          </p>
        </template>`,
      },
    });

    expect(wrapper.html()).toMatchInlineSnapshot(`

<div class="ais-StateResults">
  <p>
    Query is here
  </p>
</div>

`);
  });

  it('allows default slot to render whatever they want (falsy query)', () => {
    __setState({
      results: {
        query: '',
        hits: [{ objectID: '1', name: 'one' }, { objectID: '2', name: 'two' }],
        page: 1,
      },
      state: {},
    });

    const wrapper = mount(StateResults, {
      scopedSlots: {
        default: `
        <template slot-scope="{ query }">
          <p v-if="query">
            Query is here
          </p>
          <p v-else>
            There's no query
          </p>
        </template>`,
      },
    });

    expect(wrapper.html()).toMatchInlineSnapshot(`

<div class="ais-StateResults">
  <p>
    There's no query
  </p>
</div>

`);
  });
});
