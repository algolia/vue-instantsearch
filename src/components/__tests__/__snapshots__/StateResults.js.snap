// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders explanation if no slot is used 1`] = `

<div class="ais-StateResults">
  <p>
    Use this component to have a different layout based on a certain state.
  </p>
  <p>
    Fill in the slot, and get access to the following things on the
    <code>
      slot-scope
    </code>
    :
  </p>
  <pre>
    results: [
  "query",
  "hits",
  "page"
]
  </pre>
  <pre>
    state: [
  "query"
]
  </pre>
</div>

`;
